### DISCUSSION TITLE
Strategy to deal with TelegramRetryAfter errors

### DISCUSSION THREAD

**Discussion by 0xlucyy** [aiogram/aiogram#1489] [Q&A]
I am building a telegram gaming bot. My bot sends messages to the chat the bot operates in. The bot responds to the users in the chat. Sometimes the users spam commands to the bot. The bot is also printing required game information into the chat. 

The load on the bot described above sometimes results in the following error

```
2024-05-11 15:44:08,840 - telegram_gaming - ERROR - [errors.py:29] - Error Description: Flood control exceeded on method 'SendMessage' in chat CHAT_ID. Retry in 37 seconds.
Original description: Too Many Requests: retry after 37 in chat CHAT_ID ... Seek help: https://core.telegram.org/bots/faq#my-bot-is-hitting-limits-how-do-i-avoid-this
```

Please notice that the error messages refers to the following url for help `https://core.telegram.org/bots/faq#my-bot-is-hitting-limits-how-do-i-avoid-this`

Contained within the avoid-this url, is the following documentation on limits imposed on a bot within a particular chat group

```
When sending messages inside a particular chat, avoid sending more than one message per second. We may allow short bursts that go over this limit, but eventually you'll begin receiving 429 errors.
```

and 

```
Also note that your bot will not be able to send more than 20 messages per minute to the same group.
```

This two pieces of documentation appear at odds with each other. 

I am looking for a viable strategy to deal with this particular exception, namely `Flood control exceeded on method 'SendMessage'`. When this error message is pushed into my bot, by telegram, I cannot even send a message to the group indicating the bot has been put on timeout for X number of seconds. 

What is my recourse here? Is it possible to have 2 bots in the chat? Meaning, that if Bot_A is timed-out, I can swap Bot_A with Bot_B until Bot_A comes back online? 

I understand the reasoning of why telegram wants to impose these limits on their API. These limits present a severe obstacle for my use case and I am curious and probing for possible answers and solutions! 

Just for brevity, here is how I am handling errors, I have an `errors.py` file which is imported where the bot is init

```
from asyncio import sleep
from aiogram.exceptions import TelegramRetryAfter
from aiogram.types.error_event import ErrorEvent
from log import logger
from baseBot import dp


@dp.errors()
async def error_handler(exception):
  # Check if the exception is an ErrorEvent
  if isinstance(exception, ErrorEvent):

    if isinstance(exception.exception, TelegramRetryAfter):
      await handle_telegram_retry_after(exception)
    else:
      # Optionally handle other exceptions differently
      logger.error(f"Unhandled exception type: {type(exception.exception)}")
      logger.error(f"Unhandled exception msg: {exception.exception}")

  else:
    logger.error(f"Unhandled exception type: {type(exception).__name__}")
    logger.error(f"Unhandled exception msg: {exception}")


async def handle_telegram_retry_after(exception: TelegramRetryAfter):
  _chat_id = exception.update.message.chat.id
  _msg = exception.exception.message
  _help = exception.exception.url
  logger.error(f"Error Description: {_msg} in chat {_chat_id} ... Seek help: {_help}")
  

  # TODO NEED TO LOG THIS TO CHAT - THEN DONE
  await bot.send_message(
    _chat_id,
    f"Telegram is slowing us the fvck down for {exception.exception.retry_after} seconds ... Fvck!"
  )
  await sleep(exception.exception.retry_after)
```

**Comment by 0xlucyy** [comment]
So I ended up doing something like this for the time being

```
async def send_message_with_retry(bot, chat_id, text):
  # handle scenarios where TelegramRetryAfter may occur multiple times consecutively.
  while True:
    try:
      await bot.send_message(chat_id, text)
      return # Return from the function if the message is sent successfully
    except Exception as e:
      if isinstance(e, TelegramRetryAfter):
        _chat_id = e.method.chat_id
        _msg = e.message
        _help = e.url
        logger.warning(f"Error Description: {_msg} in chat {_chat_id} ... Seek help: {_help}")
        await sleep(e.retry_after)
        await bot.send_message(chat_id, text)
      return
```

I put that code in my errors.py file and import that function into my other source code files. I use the `send_message_with_retry` function to wrap all `bot.send_message()` 

```
    await bot.send_message(chat_id, NO_ACTIVE_GAME) # Instead of this
    await errors.send_message_with_retry(bot, chat_id, NO_ACTIVE_GAME) # Im doing this
```


While this works I MUCH prefer to simply have 2 bots to toggle between so I do not have to wait X amount of seconds. Ill wait for feedback on this thread.

**Comment by JrooTJunior** [comment]
You're getting this error because you're doing something wrong, but no one can help figure out what you're actually doing without showing your code and bot in chats.

Anyway, the code snippet provided in the comment to this question smells a bit, instead you can use the client session middleware to implement retries without changing your code in the handlers at all. [Read more >>](https://docs.aiogram.dev/en/dev-3.x/api/session/middleware.html)

**Comment by 0xlucyy** [comment]
Unfortunately your reply isnt helpful. 

It is not necessary to show my code for an error message such as `TelegramRetryAfter`.  Please note the following error message from telegram which clearly describes the issue, namely that my bot is sending too many `bot.send_message()` instructions to telegram.

`Error Description: Flood control exceeded on method 'SendMessage'`

I have the following test code to trigger a `TelegramRetryAfter` exception on method `SendMessage`
```
    done = True
    while done==True:
      done = await errors.send_message_with_retry(bot, chat_id, "testing send_message_with_retry TelegramRetryAfter wrapper")
```

My bot spams telegram messages using the following `send_message_with_retry` function. I should note the code works perfectly.
```
async def send_message_with_retry(bot, chat_id, text):
  # handle scenarios where TelegramRetryAfter may occur multiple times consecutively.
  while True:
    try:
      await bot.send_message(chat_id, text) # <----- Here is where the exception is thrown as I am spamming this method
      return True # Return from the function if the message is sent successfully
    except Exception as e:
      if isinstance(e, TelegramRetryAfter):
        _chat_id = e.method.chat_id
        _msg = e.message
        _help = e.url
        logger.warning(f"Error Description: {_msg} in chat {_chat_id} ... Seek help: {_help}")
        await sleep(e.retry_after)
        await bot.send_message(chat_id, text)
      return False
```


Here are the corresponding telegram chat logs which triggers the `TelegramRetryAfter` exception
```
User1, [May x, 2024 at x:x:x PM]:
/StartGame # <---- This triggers my test code which spams telegram with bot.send_message()

Bot_A, [May x, 2024 at x:x:x PM]:
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper

testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper

testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper

testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper
testing send_message_with_retry TelegramRetryAfter wrapper # < --- TelegramRetryAfter exception triggered here
````


What my test code does is stress test my bot to understand the limits of how many messages my bot can send into a chat. 

I need a solution for when my bot is rate limited on `bot.send_message()`. The users in my telegram chat are playing a game, and the user can issue commands to the bot. The bot outputs game information to the users. 

All these interactions with the telegram game bot sometimes trigger a `TelegramRetryAfter` on `bot.send_message()`.

What is an effective strategy to deal with this issue described above? Simply linking to a poorly documented `Client session middlewares` page does not help me. 

I was looking at `https://docs.aiogram.dev/en/latest/dispatcher/flags.html`, in particular `rate_limit()`, but the docs are poorly written.


For example, I am trying the following 

```
from aiogram import flags

@flags.rate_limit(limit=1, key='/StartGame', period=1)
@dp.message(Command('StartGame'))
async def start(message: Message, state: FSMContext):
   ....
```

In theory, this `rate_limit` flag should allow users to call `/StartGame` once per second, but the flag doesnt appear to be working because it seems the rate limit is not enforced. I am using aiogram v3.6.0 btw.

The best solution in my mind is using 2 bots, and swapping between the two as 1 is rate limited. Please advise if it is allowed to use 2 bots for this purpose?

Thank you

**Comment by 0xlucyy** [comment]
So this is what I ended up doing

```
class RetryMiddleware(BaseMiddleware):
  def __init__(self, max_retries: int = 3, delay: float = 10):
    self.max_retries = max_retries
    self.delay = delay
    self.retry_count = 0
    self.exception_count: Dict[str, int] = {}

  async def __call__(
    self,
    handler: Callable[[TelegramObject, Dict[str, Any]], Awaitable[Any]],
    event: TelegramObject,
    data: Dict[str, Any],
  ) -> Any:
    retries = 0
    if isinstance(event, Message):
      chat_id = event.chat.id
      user_id = event.from_user.id
      logger.info(f"Message text: {event.text}")
    else:
      chat_id = "Unknown"
      logger.info(f"Event type: {type(event).__name__}")

    while True:
      try:
        return await handler(event, data)
      except TelegramRetryAfter as e:
          await handle_telegram_retry_after(chat_id, e)
          await sleep(e.retry_after)
          continue  # Retry the request after sleeping
      except Exception as e:
        exception_class = e.__class__.__name__
        self.exception_count[exception_class] = self.exception_count.get(exception_class, 0) + 1
        logger.error(f'Handling {exception_class} in __call__')
        
        retries += 1
        self.retry_count += 1
        if retries >= self.max_retries:
          logger.error(f"Max retries ({self.max_retries}) reached. Raising exception...")
          raise e
        logger.warning(f"Encountered exception: {exception_class}. Retrying in {self.delay} seconds... (Retry {retries}/{self.max_retries})")
        await sleep(self.delay)
```

Since I cannot get a clear answer whether I can use multiple bots to handle a situation where a bot is deactivated due to TelegramRetryAfter, I am going to implement a solution with multiple bots.
